### DISCUSSION TITLE
Build proxy for openHAB

### DISCUSSION THREAD

**Discussion by mu88** [dotnet/yarp#2138] [Q&A] [answered]
Hi 👋🏻

I'm trying to build a reverse proxy using YARP so that I can access [openHAB](https://www.openhab.org/) via a configurable URL path segment. For example:
- openHAB is running and accessible via http://localhost:9000
- My app should proxy openHAB to http://localhost:8080/openhab

This is my current YARP config:
```c#
using Microsoft.Extensions.Primitives;
using Yarp.ReverseProxy.Configuration;
using Yarp.ReverseProxy.Transforms;

namespace OpenHabUrlSegment;

internal class OpenHabProxyConfigProvider : IProxyConfigProvider
{
    private readonly IReadOnlyList<RouteConfig> _routes;

    private readonly IReadOnlyList<ClusterConfig> _clusters;

    public OpenHabProxyConfigProvider()
    {
        _routes = new[]
        {
            new RouteConfig { RouteId = "route1", ClusterId = "cluster1", Match = new RouteMatch { Path = "openhab/{**catch-all}" } }
                .WithTransformPathRemovePrefix("/openhab")
        };

        _clusters = new[]
        {
            new ClusterConfig
            {
                ClusterId = "cluster1",
                Destinations = new Dictionary<string, DestinationConfig> { { "destination1", new DestinationConfig { Address = "http://127.0.0.1:9000/" } } }
            }
        };
    }

    public IProxyConfig GetConfig() => new OpenHabProxyConfig(_routes, _clusters);

    private class OpenHabProxyConfig : IProxyConfig
    {
        private readonly CancellationTokenSource _cts = new();

        public OpenHabProxyConfig(IReadOnlyList<RouteConfig> routes, IReadOnlyList<ClusterConfig> clusters)
        {
            Routes = routes;
            Clusters = clusters;
            ChangeToken = new CancellationChangeToken(_cts.Token);
        }

        public IReadOnlyList<RouteConfig> Routes { get; }

        public IReadOnlyList<ClusterConfig> Clusters { get; }

        public IChangeToken ChangeToken { get; }
    }
}
```

Now the following happens:
1. Navigate to http://localhost:8080/openhab in browser
2. Browser URL remains http://localhost:8080/openhab, in the logs I see that the request is proxied to http://127.0.0.1:9000.
3. The request succeeds with `HTTP 200` and the following HTML:
```html
<!doctype html>
<html>
<head>
	<meta charset="utf-8">
	<meta http-equiv="Content-Security-Policy"
		content="default-src * 'self' 'unsafe-inline' 'unsafe-eval' data: gap: content: blob:; style-src 'self' 'unsafe-inline';">
	<meta name="viewport"
		content="width=device-width,initial-scale=1,maximum-scale=1,minimum-scale=1,user-scalable=no,minimal-ui,viewport-fit=cover">
	<meta name="theme-color" content="#e64a19">
	<meta name="format-detection" content="telephone=no">
	<meta name="msapplication-tap-highlight" content="no">
	<title>openHAB</title>
	<meta name="apple-mobile-web-app-capable" content="yes">
	<meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
	<link rel="apple-touch-icon" href="/res/icons/apple-touch-icon.png" type="image/png" sizes="180x180"
		crossorigin="use-credentials">
	<link rel="icon" href="/res/icons/favicon.svg" type="image/svg+xml" sizes="any" crossorigin="use-credentials">
	<link rel="icon" href="/res/icons/128x128.png" type="image/png" sizes="128x128" crossorigin="use-credentials">
	<link rel="manifest" href="/manifest.json" crossorigin="use-credentials">
	<link href="/css/app.d702c20a4e09bf4012ea.css" rel="stylesheet">
</head>

<body>
	<div id="app"></div>
	<script src="/js/app.d702c20a4e09bf4012ea.js"></script>
</body>
</html>
```

Now the browser tries to load http://localhost:8080/js/app.d702c20a4e09bf4012ea.js which obviously fails with `HTTP 404` as the correct request would have to be http://localhost:8080/openhab/js/app.d702c20a4e09bf4012ea.js (which succeeds if I manually insert the additional URL path segment `/openhab` in the browser's address bar).

So my question is: what do I have to configure in YARP so that the correct URL including the URL path segment is used/added for all subsequent requests?

Thank you 👋🏻

PS: I've found [this GitHub issue](https://github.com/openhab/openhab-core/issues/150#issuecomment-1025099561) where an openHAB user was able to accomplish this with Apache - however, I'm struggling to "translate" this into YARP's config.

**Comment by Tratcher** [comment]
Doing a re-write based on the referrer is clever. This would have to happen in middleware before routing since routing is expecting that path prefix. Have you tried using the rewrite middleware for that?
https://learn.microsoft.com/aspnet/core/fundamentals/url-rewriting

**Reply by mu88** [reply]
No I haven't, thx for pointing me to this 💪🏻 so would I still need to use YARP or would the middleware handle everything? And what do you mean by _a re-write based on the referrer is clever_?
Thank you and have a great weekend!

**Reply by Tratcher** [reply]
You'd keep YARP as is (though you set it up in the hardest way possible 😆).
You can simplify your config using either https://microsoft.github.io/reverse-proxy/articles/config-providers.html#in-memory-config or https://microsoft.github.io/reverse-proxy/articles/config-files.html#configuration-structure

I say it's clever because fixing relative resource links is a hard problem for proxies. This is the first time I've seen the referrer used for that.

**Reply by mu88** [reply]
> You'd keep YARP as is (though you set it up in the hardest way possible 😆). You can simplify your config [...]

Thank for that hint 🙃 I chose this approach because I make use of the [options pattern](https://learn.microsoft.com/en-us/dotnet/core/extensions/options)  (missing in my sample code for brevity) and I wasn't sure how to retrieve the configured options when using minimal APIs. If I can make use of the options pattern with minimal APIs as well, I'm very happy to simplify my code.

So this is my understanding regarding the referer/URL rewrite:
1. First request is against http://localhost:8080/openhab which YARP proxies to http://localhost:9000, returning the HTML with the relative resources.
2. The browser tries to resolve the relative resources via http://localhost:8080 (e. g. http://localhost:8080/js/app.d702c20a4e09bf4012ea.js) which the URL Rewrite middleware will modify to http://localhost:8080/openhab/js/app.d702c20a4e09bf4012ea.js.
3. http://localhost:8080/openhab/js/app.d702c20a4e09bf4012ea.js will be proxied by YARP to http://localhost:9000/js/app.d702c20a4e09bf4012ea.js

Is that correct?

**Reply by mu88** [reply]
# Update
I've configured the following redirect:
```c#
app.UseRewriter(new RewriteOptions().AddRewrite("^(?!openhab/)(.*)", "openhab/$1", true));
```

Now it doesn't look as bad as before! Many resources can be loaded successfully and the browser URL remains http://localhost:8080/openhab/ 💪🏻

It looks like this (http://localhost:8080/openhab/):  
![image](https://github.com/microsoft/reverse-proxy/assets/4560672/30a51725-bc0c-4d5b-8b95-429e1f17720b)

But it should look like this (according to http://localhost:9000/):  
![image](https://github.com/microsoft/reverse-proxy/assets/4560672/7df66a56-0427-4885-ba40-af0181ee9787)

And when trying to login, the browser brings me to http://localhost:8080/auth/:  
![image](https://github.com/microsoft/reverse-proxy/assets/4560672/8ccf9c14-2c67-488a-96a8-3b6afed8f5d4)

And after entering my credentials and signing in, I end up at http://localhost:8080:
![image](https://github.com/microsoft/reverse-proxy/assets/4560672/68554575-8028-4e91-a342-c4b11b154eb0)

Seems like something is broken/missing in the auth flow 🤔

**Reply by Tratcher** [reply]
So far so good 😁 
Check the HTML for that specific resource, see if its url is in the same format as the others. Keep an eye out for minor differences like the presence of a leading '/', or the use of an absolute url.

I don't see an auth specific error there. What do you see in the Application/Cookies tab? What do the cookies look like in the POST response? Do they have a path?

**Reply by samsp-msft** [reply]
The referrer hack is probably broken with auth as it is going to be coming from the auth page, not a /openhab/ url.

**Reply by Tratcher** [reply]
Ah, in that case the return_url could be fixed up in the Location header on the prior response.

**Reply by mu88** [reply]
I don't have the code in front of me yet, but I remember that the auth URL is built up in a JavaScript file using `window.location.origin` - something like `window.location.origin + "/auth"`. How does the browser evaluate `window.location.origin`?

**Reply by mu88** [reply]
Do you guys have any further ideas, @Tratcher / @samsp-msft? I'd really appreciate it

**Reply by Tratcher** [reply]
Look at the response from the initial `/auth` request, there should be a Location header with a redirect_url there. You should be able to use a yarp response transform to re-write that.

**Reply by mu88** [reply]
Sry for opening this again. I came back to this today and there seems to be a change in the codebase of openHAB: the first call now returns an HTTP 302 with the `location` header pointing to http://localhost:9000/, thereby sending the browser to http://localhost:9000/ and leaving my proxy site http://localhost:8080/openhab:
```
info: Yarp.ReverseProxy.Forwarder.HttpForwarder[9]
      Proxying to http://127.0.0.1:9000/openhab HTTP/2 RequestVersionOrLower no-streaming
info: Yarp.ReverseProxy.Forwarder.HttpForwarder[56]
      Received HTTP/1.1 response 302.
```

How can I stop YARP from sending this `HTTP 302` to the client?

**Reply by samsp-msft** [reply, marked as answer]
It seems like OpenHab is not well setup to be proxied. At this point it may be more efficient to submit changes there so that it honors the x-forward headers when its generating URL links. The OpenHab community are pretty open and friendly.

I haven't looked at how the configuration UI site is written, is it in Java like the bindings and runtime infrastructure? My house runs on OpenHab - I have an hdmi-cec binding I wrote to enable control of home theatre devices that don't have their own bindings.

**Reply by mu88** [reply]
Seems to be all Java, yes
